ext/drreg/drreg.c

```diff
@@ -1719,35 +2549,62 @@ drreg_event_restore_state(void *drcontext, bool restore_memory,
         return true; /* fault not in cache */
     for (reg = DR_REG_START_GPR; reg <= DR_REG_STOP_GPR; reg++)
         spilled_to[GPR_IDX(reg)] = MAX_SPILLS;
+    for (reg = DR_REG_APPLICABLE_START_SIMD; reg <= DR_REG_APPLICABLE_STOP_SIMD; reg++)
+        spilled_simd_to[SIMD_IDX(reg)] = MAX_SIMD_SPILLS;
+
     LOG(drcontext, DR_LOG_ALL, 3,
         "%s: processing fault @" PFX ": decoding from " PFX "\n", __FUNCTION__,
         info->raw_mcontext->pc, pc);
     instr_init(drcontext, &inst);
+    instr_init(drcontext, &next_inst);
     while (pc < info->raw_mcontext->pc) {
         instr_reset(drcontext, &inst);
+        instr_reset(drcontext, &next_inst);
         prev_pc = pc;
         pc = decode(drcontext, pc, &inst);
+        decode(drcontext, pc, &next_inst);
 
-        if (is_our_spill_or_restore(drcontext, &inst, &spill, &reg, &slot, &offs)) {
+        if (is_our_spill_or_restore(drcontext, &inst, &next_inst, &is_spill, &reg, &slot,
+                                    &offs, &is_indirect_spill)) {
             LOG(drcontext, DR_LOG_ALL, 3,
                 "%s @" PFX " found %s to %s offs=0x%x => slot %d\n", __FUNCTION__,
-                prev_pc, spill ? "spill" : "restore", get_register_name(reg), offs, slot);
-            if (spill) {
-                if (slot == AFLAGS_SLOT) {
+                prev_pc, is_spill ? "is_spill" : "restore", get_register_name(reg), offs,
+                slot);
+            if (is_spill) {
+                if (is_indirect_spill) {
+                    if (reg_is_vector_simd(reg)) {
+                        if (spilled_simd_to[SIMD_IDX(reg)] < MAX_SIMD_SPILLS &&
+                            /* allow redundant spill */
+                            spilled_simd_to[SIMD_IDX(reg)] != slot) {
+                            /* This reg is already spilled: we assume that this new spill
+                             * is to a tmp slot for preserving the tool's value.
+                             */
+                            LOG(drcontext, DR_LOG_ALL, 3,
+                                "%s @" PFX ": ignoring tool is_spill\n", __FUNCTION__,
+                                pc);
+                        } else {
+                            spilled_simd_to[SIMD_IDX(reg)] = slot;
```

Comment:
Info about whether xmm, ymm, or zmm lost?

Reply:
(Fixed this, wait for my update) Still not sure if we should do xmm-only first to limit the patch size